libs/features/jarvis-chat.js

```diff
@@ -0,0 +1,344 @@
+        + `${data['event.type'] || ''}`,
+      eventAction: `${data['event.subcategory'] || ''}:`
+        + `${data['content.name'] || ''}:`
+        + `${data['event.subtype'] || ''}:`
+        + `${data['event.type'] || ''}`,
+    },
+  };
+
+  window.digitalData.chat = { chatInfo: { primaryProduct: { productName: data['event.subtype'] } } };
+
+  sendEvent(data);
+};
+
+const sendSurveyFeedbackEvent = (data) => {
+  window.digitalData.primaryEvent = {
+    eventInfo: {
+      eventName: 'chat:'
+        + `${data['event.workflow'] || ''}:`
+        + `${data['content.name'] || ''}:`
+        + `${data['event.subtype'] || ''}:`
+        + `${data['event.type'] || ''}:`
+        + `${data['content.id'] || ''}`,
+      eventAction: `${data['event.subcategory'] || ''}:`
+        + `${data['content.name'] || ''}:`
+        + `${data['event.subtype'] || ''}:`
+        + `${data['event.type'] || ''}`,
+    },
+  };
+
+  sendEvent(data);
+};
+
+const sendChatErrorEvent = (data) => {
+  window.digitalData.primaryEvent = {
+    eventInfo: {
+      eventName: 'chat:'
+        + `${data['event.workflow'] || ''}:`
+        + `${data['event.subtype'] || ''}:`
+        + `${data['event.type'] || ''}:`
+        + 'error',
+      eventAction: `${data['event.subcategory'] || ''}:`
+        + `${data['event.subtype'] || ''}:`
+        + `${data['event.type'] || ''}:`
+        + 'error',
+    },
+  };
+
+  window.digitalData.chat = {};
+  window.digitalData.chat.chatInfo = getBaseEvent(data);
+  window.digitalData.chat.chatInfo.chatErrorCode = data['event.error_code'];
+  window.digitalData.chat.chatInfo.chatErrorType = data['event.error_type'];
+  window.digitalData.chat.chatInfo.chatErrorDescription = data['event.error_desc'];
+
+  sendEvent(data);
+};
+
+const sendPrimaryEvent = (data) => {
+  window.digitalData.primaryEvent = {
+    eventInfo: {
+      eventName: 'chat:'
+        + `${data['event.workflow'] || ''}:`
+        + `${data['content.name'] || ''}:`
+        + `${data['event.subtype'] || ''}:`
+        + `${data['event.type'] || ''}`,
+      eventAction: `${data['event.subcategory'] || ''}:`
+        + `${data['content.name'] || ''}:`
+        + `${data['event.subtype'] || ''}:`
+        + `${data['event.type'] || ''}`,
+    },
+  };
+
+  window.digitalData.chat = {};
+  window.digitalData.chat.chatInfo = getBaseEvent(data);
+
+  sendEvent(data);
+};
+
+const redirectToSupport = () => window.location.assign('https://helpx.adobe.com');
+
+const isChatOpen = () => {
+  const instance = window.AdobeMessagingExperienceClient;
+  return instance?.isAdobeMessagingClientInitialized()
+    && instance?.getMessagingExperienceState()?.windowState !== 'hidden';
+};
+
+const openChat = (event) => {
+  if (!chatInitialized) redirectToSupport();
+  const open = window.AdobeMessagingExperienceClient?.openMessagingWindow;
+  if (!open) return;
+  if (isChatOpen()) return;
+  if (event) {
+    const sourceType = event?.target?.tagName?.toLowerCase();
```

Comment:
You already have a check if event exists. Target should always exist and should be a htmlElement you don't need any of the three checks.

Looks overly defensive to me, but there's no real downside

Reply:
You're correct, I can update. However, there are use-cases when the event is not set.